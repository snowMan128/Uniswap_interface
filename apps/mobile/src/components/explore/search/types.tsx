import { SearchResult } from 'wallet/src/features/search/SearchResult'
import { SEARCH_RESULT_HEADER_KEY } from './constants'

// Header type used to render header text instead of SearchResult item

export type SearchResultOrHeader =
  | SearchResult
  | { type: typeof SEARCH_RESULT_HEADER_KEY; title: string }
