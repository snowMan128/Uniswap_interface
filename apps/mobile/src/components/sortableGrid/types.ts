import { FlatList, ScrollView, View } from 'react-native'
import { SharedValue } from 'react-native-reanimated'

export type Require<T, K extends keyof T = keyof T> = Required<Pick<T, K>> & Omit<T, K>

export type ItemMeasurements = {
  height: number
  width: number
  x: number
  y: number
}

export type AutoScrollProps = {
  scrollableRef: React.RefObject<FlatList | ScrollView>
  visibleHeight: SharedValue<number>
  scrollY: SharedValue<number>
  // The parent container inside the scrollable that wraps the grid
  // (e.g. when the grid is rendered inside the FlatList header)
  // if not provided, we assume that the grid is the first child in
  // the scrollable container
  containerRef?: React.RefObject<View>
}

export type SortableGridContextType = {
  gridContainerRef: React.RefObject<View>
  itemAtIndexMeasurements: SharedValue<ItemMeasurements[]>
  dragActivationProgress: SharedValue<number>
  activeIndex: number | null
  previousActiveIndex: SharedValue<number | null>
  activeTranslation: SharedValue<{ x: number; y: number }>
  scrollOffsetDiff: SharedValue<number>
  renderIndexToDisplayIndex: SharedValue<number[]>
  setActiveIndex: (index: number | null) => void
  onDragStart?: () => void
  displayToRenderIndex: SharedValue<number[]>
  activeItemScale: SharedValue<number>
  visibleHeight: SharedValue<number>
  activeItemOpacity: SharedValue<number>
  activeItemShadowOpacity: SharedValue<number>
  touchedIndex: SharedValue<number | null>
  editable: boolean
  containerStartOffset: SharedValue<number>
  containerEndOffset: SharedValue<number>
}

export type SortableGridRenderItemInfo<I> = {
  item: I
  index: number
  dragActivationProgress: SharedValue<number>
  isTouched: SharedValue<boolean>
}

export type SortableGridRenderItem<I> = (info: SortableGridRenderItemInfo<I>) => JSX.Element

export type Vector = {
  x: number
  y: number
}

export type SortableGridChangeEvent<I> = {
  data: I[]
  fromIndex: number
  toIndex: number
}
