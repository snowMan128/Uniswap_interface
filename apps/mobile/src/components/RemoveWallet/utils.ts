import { CommonActions } from '@react-navigation/core'
import { dispatchNavigationAction } from 'src/app/navigation/rootNavigation'
import { OnboardingScreens, Screens } from 'src/screens/Screens'
import { ImportType, OnboardingEntryPoint } from 'wallet/src/features/onboarding/types'

// This fast-forwards user to the same app state as if
// they have pressed "Get Started" on Landing and should now see import method view
export function navigateToOnboardingImportMethod(): void {
  dispatchNavigationAction(
    CommonActions.reset({
      index: 0,
      routes: [
        {
          name: Screens.OnboardingStack,
          state: {
            index: 1,
            routes: [
              {
                name: OnboardingScreens.Landing,
                params: {
                  entryPoint: OnboardingEntryPoint.FreshInstallOrReplace,
                  importType: ImportType.NotYetSelected,
                },
              },
              {
                name: OnboardingScreens.ImportMethod,
                params: {
                  entryPoint: OnboardingEntryPoint.FreshInstallOrReplace,
                  importType: ImportType.NotYetSelected,
                },
              },
            ],
          },
        },
      ],
    })
  )
}
