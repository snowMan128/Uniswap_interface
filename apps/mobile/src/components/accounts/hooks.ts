import { NetworkStatus, WatchQueryFetchPolicy } from '@apollo/client'
import {
  AccountListQuery,
  // eslint-disable-next-line no-restricted-imports
  useAccountListQuery,
} from 'wallet/src/data/__generated__/types-and-hooks'
// eslint-disable-next-line no-restricted-imports
import { usePortfolioValueModifiers } from 'wallet/src/features/dataApi/balances'
import { GqlResult } from 'wallet/src/features/dataApi/types'

export function useAccountList({
  addresses,
  fetchPolicy,
  notifyOnNetworkStatusChange,
}: {
  addresses: Address | Address[]
  fetchPolicy?: WatchQueryFetchPolicy
  notifyOnNetworkStatusChange?: boolean | undefined
}): GqlResult<AccountListQuery> & {
  startPolling: (pollInterval: number) => void
  stopPolling: () => void
  networkStatus: NetworkStatus
  refetch: () => void
} {
  const valueModifiers = usePortfolioValueModifiers(addresses)
  const { data, loading, networkStatus, refetch, startPolling, stopPolling } = useAccountListQuery({
    variables: { addresses, valueModifiers },
    notifyOnNetworkStatusChange,
    fetchPolicy,
  })

  return {
    data,
    loading,
    networkStatus,
    refetch,
    startPolling,
    stopPolling,
  }
}
