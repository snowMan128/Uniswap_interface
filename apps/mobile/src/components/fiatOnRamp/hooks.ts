import { useIsDarkMode } from 'ui/src'
import { FORLogo } from 'wallet/src/features/fiatOnRamp/types'

export function useFiatOnRampLogoUrl(logos: FORLogo | undefined): string | undefined {
  const isDarkMode = useIsDarkMode()

  if (!logos) {
    return
  }

  return isDarkMode ? logos.darkLogo : logos.lightLogo
}
