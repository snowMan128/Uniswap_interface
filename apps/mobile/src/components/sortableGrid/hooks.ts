import { useCallback, useRef } from 'react'
import { FlatList, ScrollView } from 'react-native'
import { runOnJS, SharedValue, useAnimatedReaction, useSharedValue } from 'react-native-reanimated'
import { AUTO_SCROLL_THRESHOLD } from './constants'
import { useSortableGridContext } from './SortableGridProvider'
import { ItemMeasurements } from './types'

export function useStableCallback<
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  C extends (...args: Array<any>) => any
>(callback?: C): C {
  const callbackRef = useRef(callback)
  callbackRef.current = callback

  return useCallback(
    // eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/no-unsafe-return
    (...args: Array<any>) => callbackRef.current?.(...args),
    []
  ) as C
}

export function useAnimatedZIndex(renderIndex: number): SharedValue<number> {
  const { touchedIndex: touchedIndexValue, previousActiveIndex: previousActiveIndexValue } =
    useSortableGridContext()
  const zIndexValue = useSharedValue(0)

  useAnimatedReaction(
    () => ({
      touchedIndex: touchedIndexValue.value,
      previousActiveIndex: previousActiveIndexValue.value,
    }),
    ({ touchedIndex, previousActiveIndex }) => {
      if (renderIndex === touchedIndex) {
        // Display the currently touched item on top of all other items
        zIndexValue.value = 10000
      } else if (renderIndex === previousActiveIndex) {
        // Display the previously active item on top of other items
        // except the currently touched item (used to properly position
        // items before their drop animations finishes)
        zIndexValue.value = 9999
      } else {
        zIndexValue.value = 0
      }
    }
  )

  return zIndexValue
}

export function useItemOrderUpdater(
  renderIndex: number,
  activeRenderIndex: number | null,
  displayIndexValue: SharedValue<number>,
  numColumns: number
): void {
  const {
    touchedIndex: touchedIndexValue,
    activeTranslation: activeTranslationValue,
    itemAtIndexMeasurements: itemAtIndexMeasurementsValue,
    displayToRenderIndex: displayToRenderIndexValue,
    renderIndexToDisplayIndex: renderIndexToDisplayIndexValue,
    scrollOffsetDiff: scrollOffsetDiffValue,
  } = useSortableGridContext()

  useAnimatedReaction(
    () => ({
      activeTranslation: activeTranslationValue.value,
      displayIndex: displayIndexValue.value,
      scrollOffsetDiff: scrollOffsetDiffValue.value,
      touchedIndex: touchedIndexValue.value,
    }),
    ({ displayIndex, activeTranslation, scrollOffsetDiff, touchedIndex }) => {
      // Return if there is no active item or the current item is the active item
      // (only active item neighbors decide if the order should be updated)
      if (
        activeRenderIndex === null ||
        touchedIndex === null ||
        activeRenderIndex === renderIndex
      ) {
        return
      }

      const itemAtIndexMeasurements = itemAtIndexMeasurementsValue.value
      // The current item might have moved so we get its measurements based on
      // the display index (e.g. if the item that is at index 0 in the data array
      // was moved to index 1, its current offset is determined by the cell that
      // is at index 1 in the grid)
      const itemMeasurements = itemAtIndexMeasurements[displayIndex]
      // For the active item, we always use the render index to get the measurements
      // as the activeTranslation is calculated in relation to the render position
      const activeItemMeasurements = itemAtIndexMeasurements[activeRenderIndex]
      if (!itemMeasurements || !activeItemMeasurements) {
        return
      }

      // Active item data
      const activeAbsoluteX = activeItemMeasurements.x + activeTranslation.x
      const activeAbsoluteY = activeItemMeasurements.y + activeTranslation.y
      const activeWidth = activeItemMeasurements.width
      const activeHeight = activeItemMeasurements.height

      // Current item data
      const itemAbsoluteX = itemMeasurements.x
      const itemAbsoluteY = itemMeasurements.y - scrollOffsetDiff
      const itemWidth = itemMeasurements.width
      const itemHeight = itemMeasurements.height

      //
      // Check if the current element is on the boundary of the container
      //
      const renderIndexToDisplayIndex = renderIndexToDisplayIndexValue.value

      const itemsCount = renderIndexToDisplayIndex.length
      const columnIndex = displayIndex % numColumns
      const rowIndex = Math.floor(displayIndex / numColumns)
      const itemsInColumnCount =
        Math.floor(itemsCount / numColumns) + (columnIndex < itemsCount % numColumns ? 1 : 0)

      const isInFirstRow = displayIndex < numColumns
      const isLastInColumn = rowIndex === itemsInColumnCount - 1
      const isInLastColumn = columnIndex === numColumns - 1
      const isInFirstColumn = columnIndex === 0

      // Return if the active item is not overlapping the current item
      // by at least 50% in any direction (with the exception of the boundary items)
      if (
        // Right neighbor of the active item
        (activeAbsoluteX + activeWidth < itemAbsoluteX + itemWidth / 2 && !isInFirstColumn) ||
        // Left neighbor of the active item
        (activeAbsoluteX > itemAbsoluteX + itemWidth / 2 && !isInLastColumn) ||
        // Bottom neighbor of the active item
        (activeAbsoluteY + activeHeight < itemAbsoluteY + itemHeight / 2 && !isInFirstRow) ||
        // Top neighbor of the active item
        (activeAbsoluteY > itemAbsoluteY + itemHeight / 2 && !isLastInColumn)
      ) {
        return
      }

      const displayToRenderIndex = displayToRenderIndexValue.value
      const activeDisplayIndex = renderIndexToDisplayIndex[activeRenderIndex]
      if (activeDisplayIndex === undefined) {
        return
      }

      //
      // Swap the order of the current item and the active item
      //
      if (displayIndex < activeDisplayIndex) {
        // Insert the current item before the active item
        displayToRenderIndexValue.value = [
          ...displayToRenderIndex.slice(0, displayIndex),
          activeRenderIndex,
          ...displayToRenderIndex.slice(displayIndex, activeDisplayIndex),
          ...displayToRenderIndex.slice(activeDisplayIndex + 1),
        ]
      } else {
        // Insert the current item after the active item
        displayToRenderIndexValue.value = [
          ...displayToRenderIndex.slice(0, activeDisplayIndex),
          ...displayToRenderIndex.slice(activeDisplayIndex + 1, displayIndex + 1),
          activeRenderIndex,
          ...displayToRenderIndex.slice(displayIndex + 1),
        ]
      }
    },
    [renderIndex, activeRenderIndex]
  )
}

export function useAutoScroll(
  activeIndex: number | null,
  touchedIndex: SharedValue<number | null>,
  itemAtIndexMeasurements: SharedValue<Record<number, ItemMeasurements>>,
  activeTranslation: SharedValue<{ x: number; y: number }>,
  scrollOffsetDiff: SharedValue<number>,
  containerStartOffset: SharedValue<number>,
  containerEndOffset: SharedValue<number>,
  visibleHeightValue: SharedValue<number>,
  scrollYValue: SharedValue<number>,
  scrollableRef: React.RefObject<ScrollView | FlatList>
): void {
  const scrollTarget = useSharedValue(0)
  const scrollDirection = useSharedValue(0) // 1 = down, -1 = up

  const scrollToOffset = useStableCallback((offset: number) => {
    const scrollable = scrollableRef.current
    if (!scrollable || touchedIndex.value === null) {
      return
    }

    if ('scrollTo' in scrollable) {
      scrollable.scrollTo({ y: offset, animated: true })
    } else {
      scrollable.scrollToOffset({ offset, animated: true })
    }
  })

  useAnimatedReaction(
    () => touchedIndex.value,
    () => {
      // Reset when the active index changes
      scrollDirection.value = 0
    }
  )

  useAnimatedReaction(
    () => {
      if (activeIndex === null) {
        return null
      }
      const activeMeasurements = itemAtIndexMeasurements.value[activeIndex]
      if (!activeMeasurements) {
        return null
      }

      return {
        itemAbsoluteY:
          activeMeasurements.y +
          activeTranslation.value.y +
          containerStartOffset.value +
          scrollOffsetDiff.value,
        activeHeight: activeMeasurements.height,
        minOffset: containerStartOffset.value,
        maxOffset: containerEndOffset.value - visibleHeightValue.value,
        visibleHeight: visibleHeightValue.value,
        scrollY: scrollYValue.value,
      }
    },
    (props) => {
      if (!props) {
        return
      }
      const { itemAbsoluteY, scrollY, minOffset, maxOffset, activeHeight, visibleHeight } = props

      let currentScrollTarget = scrollTarget.value
      let currentScrollDirection = scrollDirection.value

      /**
       * |----------------------|
       * | content above grid   |
       * |----------------------| <- minOffset (- threshold to scroll a bit above the grid)
       * | invisible grid above |
       * | (optional)           | - if the scrollable container was scrolled down
       * |----------------------|
       * | visible grid part    |
       * |----------------------|
       * | invisible grid below | - if the scrollable container was scrolled up enough
       * | (optional)           |
       * |----------------------| <- maxOffset (+ threshold to scroll a bit below the grid)
       * | content below grid   |
       * |----------------------|
       */
      // If the active item is above the current scroll position (with small threshold
      // to start scrolling earlier) and the scroll position is not at the top of the
      // grid, scroll up
      if (
        itemAbsoluteY < scrollY + AUTO_SCROLL_THRESHOLD &&
        scrollY > minOffset - AUTO_SCROLL_THRESHOLD
      ) {
        currentScrollTarget = Math.max(minOffset - AUTO_SCROLL_THRESHOLD, scrollY - activeHeight)
        currentScrollDirection = -1
      }
      // If the active item is below the current scroll position (with small threshold
      // to start scrolling earlier) and the scroll position is not at the bottom of the
      // grid, scroll down
      else if (
        itemAbsoluteY + activeHeight > scrollY + visibleHeight - AUTO_SCROLL_THRESHOLD &&
        scrollY < maxOffset + AUTO_SCROLL_THRESHOLD
      ) {
        currentScrollTarget = Math.min(maxOffset + AUTO_SCROLL_THRESHOLD, scrollY + activeHeight)
        currentScrollDirection = 1
      }

      const scrollDiff = Math.abs(currentScrollTarget - scrollTarget.value)

      if (
        // Don't scroll if the difference is too small (limit JS thread updates that
        // become laggy when too many are triggered) and the scroll direction is the same
        // as before and the scroll target is still far enough from the min/max offset
        (scrollDiff < 0.75 * activeHeight &&
          currentScrollDirection === scrollDirection.value &&
          currentScrollTarget > minOffset - AUTO_SCROLL_THRESHOLD &&
          currentScrollTarget < maxOffset + AUTO_SCROLL_THRESHOLD) ||
        // Don't scroll if the difference is too small and the target can be considered
        // reached
        Math.abs(scrollDiff) < 2
      ) {
        return
      }

      scrollDirection.value = currentScrollDirection
      scrollTarget.value = currentScrollTarget
      runOnJS(scrollToOffset)(currentScrollTarget)
    },
    [activeIndex]
  )
}
