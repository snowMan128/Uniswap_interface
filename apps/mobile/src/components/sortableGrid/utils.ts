import { FlatList, ScrollView } from 'react-native'

const hasProp = <O extends object, P extends string>(
  object: O,
  prop: P
): object is O & Record<P, unknown> => {
  return prop in object
}

export const defaultKeyExtractor = <I>(item: I, index: number): string => {
  if (typeof item === 'string') {
    return item
  }

  if (typeof item === 'object' && item !== null) {
    if (hasProp(item, 'id')) {
      return String(item.id)
    }
    if (hasProp(item, 'key')) {
      return String(item.key)
    }
  }

  return String(index)
}

export const isScrollView = (scrollable: ScrollView | FlatList): scrollable is ScrollView => {
  return 'scrollTo' in scrollable
}
