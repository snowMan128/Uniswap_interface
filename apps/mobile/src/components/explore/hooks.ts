import { SharedEventName } from '@uniswap/analytics-events'
import { useCallback, useMemo } from 'react'
import { useTranslation } from 'react-i18next'
import { NativeSyntheticEvent, Share } from 'react-native'
import { ContextMenuAction, ContextMenuOnPressNativeEvent } from 'react-native-context-menu-view'
import { useSelectHasTokenFavorited, useToggleFavoriteCallback } from 'src/features/favorites/hooks'
import { openModal } from 'src/features/modals/modalSlice'
import { sendMobileAnalyticsEvent } from 'src/features/telemetry'
import { MobileEventName, ShareableEntity } from 'src/features/telemetry/constants'
import { logger } from 'utilities/src/logger/logger'
import { ChainId } from 'wallet/src/constants/chains'
import { AssetType } from 'wallet/src/entities/assets'
import { useCopyTokenAddressCallback } from 'wallet/src/features/tokens/hooks'
import {
  CurrencyField,
  TransactionState,
} from 'wallet/src/features/transactions/transactionState/types'
import { useAppDispatch } from 'wallet/src/state'
import { ElementName, ModalName, SectionNameType } from 'wallet/src/telemetry/constants'
import { getTokenUrl } from 'wallet/src/utils/linking'

import { CurrencyId, currencyIdToAddress } from 'wallet/src/utils/currencyId'

interface TokenMenuParams {
  currencyId: CurrencyId
  chainId: ChainId
  analyticsSection: SectionNameType
  // token, which are in favorite section would have it defined
  onEditFavorites?: () => void
}

// Provide context menu related data for token
export function useExploreTokenContextMenu({
  currencyId,
  chainId,
  analyticsSection,
  onEditFavorites,
}: TokenMenuParams): {
  menuActions: Array<ContextMenuAction & { onPress: () => void }>
  onContextMenuPress: (e: NativeSyntheticEvent<ContextMenuOnPressNativeEvent>) => void
} {
  const { t } = useTranslation()
  const isFavorited = useSelectHasTokenFavorited(currencyId)
  const dispatch = useAppDispatch()

  // `address` is undefined for native currencies, so we want to extract it from
  // currencyId, where we have hardcoded addresses for native currencies
  const currencyAddress = currencyIdToAddress(currencyId)

  const onPressCopyContractAddress = useCopyTokenAddressCallback(currencyAddress)

  const onPressShare = useCallback(async () => {
    const tokenUrl = getTokenUrl(currencyId)
    if (!tokenUrl) {
      return
    }
    try {
      await Share.share({
        message: tokenUrl,
      })
      sendMobileAnalyticsEvent(MobileEventName.ShareButtonClicked, {
        entity: ShareableEntity.Token,
        url: tokenUrl,
      })
    } catch (error) {
      logger.error(error, { tags: { file: 'balances/hooks.ts', function: 'onPressShare' } })
    }
  }, [currencyId])

  const toggleFavoriteToken = useToggleFavoriteCallback(currencyId, isFavorited)

  const onPressSwap = useCallback(() => {
    const swapFormState: TransactionState = {
      exactCurrencyField: CurrencyField.INPUT,
      exactAmountToken: '0',
      [CurrencyField.INPUT]: null,
      [CurrencyField.OUTPUT]: {
        chainId,
        address: currencyAddress,
        type: AssetType.Currency,
      },
    }
    dispatch(openModal({ name: ModalName.Swap, initialState: swapFormState }))
    sendMobileAnalyticsEvent(SharedEventName.ELEMENT_CLICKED, {
      element: ElementName.Swap,
      section: analyticsSection,
    })
  }, [analyticsSection, chainId, currencyAddress, dispatch])

  const onPressToggleFavorite = useCallback(() => {
    toggleFavoriteToken()
  }, [toggleFavoriteToken])

  const menuActions = useMemo(
    () => [
      {
        title: isFavorited ? t('Remove favorite') : t('Favorite token'),
        systemIcon: isFavorited ? 'heart.fill' : 'heart',
        onPress: onPressToggleFavorite,
      },
      ...(onEditFavorites
        ? [
            {
              title: t('Edit favorites'),
              systemIcon: 'square.and.pencil',
              onPress: onEditFavorites,
            },
          ]
        : []),
      { title: t('Swap'), systemIcon: 'arrow.2.squarepath', onPress: onPressSwap },
      {
        title: t('Copy contract address'),
        systemIcon: 'doc.on.doc',
        onPress: onPressCopyContractAddress,
      },
      ...(!onEditFavorites
        ? [
            {
              title: t('Share'),
              systemIcon: 'square.and.arrow.up',
              onPress: onPressShare,
            },
          ]
        : []),
    ],
    [
      isFavorited,
      t,
      onPressToggleFavorite,
      onEditFavorites,
      onPressSwap,
      onPressCopyContractAddress,
      onPressShare,
    ]
  )

  const onContextMenuPress = useCallback(
    async (e: NativeSyntheticEvent<ContextMenuOnPressNativeEvent>): Promise<void> => {
      await menuActions[e.nativeEvent.index]?.onPress?.()
    },
    [menuActions]
  )

  return { menuActions, onContextMenuPress }
}
