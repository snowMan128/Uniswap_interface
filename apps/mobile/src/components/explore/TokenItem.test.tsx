import * as tokenDetailsHooks from 'src/components/TokenDetails/hooks'
import { TOKEN_ITEM_DATA, tokenItemData } from 'src/test/fixtures'
import { fireEvent, render, within } from 'src/test/test-utils'
import { TokenMetadataDisplayType } from 'wallet/src/features/wallet/types'
import { ON_PRESS_EVENT_PAYLOAD } from 'wallet/src/test/fixtures'
import { buildCurrencyId } from 'wallet/src/utils/currencyId'
import { TokenItem } from './TokenItem'
import * as exploreHooks from './hooks'

describe('TokenItem', () => {
  const mockedTokenDetailsNavigation = {
    navigate: jest.fn(),
    navigateWithPop: jest.fn(),
    preload: jest.fn(),
  }

  beforeAll(() => {
    jest
      .spyOn(tokenDetailsHooks, 'useTokenDetailsNavigation')
      .mockReturnValue(mockedTokenDetailsNavigation)
    jest.spyOn(exploreHooks, 'useExploreTokenContextMenu').mockReturnValue({
      menuActions: [],
      onContextMenuPress: jest.fn(),
    })
  })

  it('renders without error', () => {
    const tree = render(<TokenItem index={0} tokenItemData={TOKEN_ITEM_DATA} />)

    expect(tree).toMatchSnapshot()
  })

  it('renders correct token number based on index', () => {
    const data = tokenItemData()
    const { queryByText } = render(<TokenItem index={1} tokenItemData={data} />)

    expect(queryByText('2')).toBeTruthy()
  })

  it('renders proper token name', () => {
    const data = tokenItemData()
    const { queryByText } = render(<TokenItem index={0} tokenItemData={data} />)

    expect(queryByText(data.name)).toBeTruthy()
  })

  it('navigates to the token details screen when pressed', () => {
    const data = tokenItemData()
    const { getByTestId } = render(<TokenItem index={0} tokenItemData={data} />)

    fireEvent.press(getByTestId(`token-item-${data.name}`), ON_PRESS_EVENT_PAYLOAD)

    expect(mockedTokenDetailsNavigation.navigate).toHaveBeenCalledWith(
      buildCurrencyId(data.chainId, data.address)
    )
  })

  describe('token price', () => {
    it('renders token price if it is provided', () => {
      const data = tokenItemData({ price: 123.45 })
      const { getByTestId } = render(<TokenItem index={0} tokenItemData={data} />)

      const tokenPrice = getByTestId('token-item/price')

      expect(within(tokenPrice).queryByText('$123.45')).toBeTruthy()
      expect(within(tokenPrice).queryByText('-')).toBeFalsy()
    })

    it('renders price placeholder if token price is not provided', () => {
      const data = tokenItemData({ price: undefined })
      const { getByTestId } = render(<TokenItem index={0} tokenItemData={data} />)

      const tokenPrice = getByTestId('token-item/price')

      expect(within(tokenPrice).queryByText('-')).toBeTruthy()
    })
  })

  describe('token price change', () => {
    it('renders token price change if it is provided', () => {
      const data = tokenItemData({ pricePercentChange24h: 12.34 })
      const { getByTestId } = render(<TokenItem index={0} tokenItemData={data} />)

      const relativeChange = getByTestId('relative-change')

      expect(within(relativeChange).queryByText('12.34%')).toBeTruthy()
    })

    it('renders price change placeholder if token price change is not provided', () => {
      const data = tokenItemData({ pricePercentChange24h: undefined })
      const { getByTestId } = render(<TokenItem index={0} tokenItemData={data} />)

      const relativeChange = getByTestId('relative-change')

      expect(within(relativeChange).queryByText('-')).toBeTruthy()
    })
  })

  describe('metadata subtitle', () => {
    const data = tokenItemData({
      marketCap: 123.45,
      volume24h: 234.56,
      totalValueLocked: 345.67,
    })

    const cases = [
      { test: 'market cap', type: TokenMetadataDisplayType.MarketCap, expected: '$123.45 MCap' },
      { test: 'volume', type: TokenMetadataDisplayType.Volume, expected: '$234.56 Vol' },
      { test: 'total value locked', type: TokenMetadataDisplayType.TVL, expected: '$345.67 TVL' },
      { test: 'symbol', type: TokenMetadataDisplayType.Symbol, expected: data.symbol },
    ]

    it.each(cases)('renders $test metadata subtitle', ({ type, expected }) => {
      const { getByTestId } = render(
        <TokenItem index={0} metadataDisplayType={type} tokenItemData={data} />
      )

      const metadataSubtitle = getByTestId('token-item/metadata-subtitle')

      expect(within(metadataSubtitle).queryByText(expected)).toBeTruthy()
    })
  })
})
