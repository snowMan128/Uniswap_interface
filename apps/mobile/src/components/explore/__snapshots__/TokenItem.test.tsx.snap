// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TokenItem renders without error 1`] = `
<View
  actions={[]}
  onPress={[MockFunction]}
>
  <View
    cancelable={true}
    disabled={false}
    focusable={true}
    hitSlop={[Function]}
    minPressDuration={0}
    onBlur={[Function]}
    onFocus={[Function]}
    onLongPress={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onPress={[Function]}
    onPressIn={[Function]}
    onPressOut={[Function]}
    style={
      {
        "flexDirection": "column",
        "opacity": 1,
        "transform": [
          {
            "scale": 1,
          },
        ],
      }
    }
    testID="token-item-cum"
  >
    <View
      animatedStyle={
        {
          "value": {},
        }
      }
      collapsable={false}
      style={
        {
          "flexDirection": "row",
          "flexGrow": 1,
          "gap": 12,
          "paddingBottom": 8,
          "paddingLeft": 24,
          "paddingRight": 24,
          "paddingTop": 8,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "flexDirection": "row",
            "gap": 4,
            "justifyContent": "center",
            "overflow": "hidden",
          }
        }
      >
        <View
          style={
            {
              "flexDirection": "column",
              "minWidth": 16,
            }
          }
        >
          <Text
            allowFontScaling={true}
            maxFontSizeMultiplier={1.2}
            style={
              {
                "color": "#7D7D7D",
                "fontFamily": "Basel-Medium",
                "fontSize": 15,
                "fontWeight": "500",
                "lineHeight": 16,
              }
            }
            suppressHighlighting={true}
          >
            1
          </Text>
        </View>
        <View
          style={
            {
              "alignItems": "center",
              "flexDirection": "column",
              "height": 40,
              "justifyContent": "center",
              "width": 40,
            }
          }
        >
          <Image
            source={
              {
                "uri": "https://loremflickr.com/640/480",
              }
            }
            style={
              [
                {
                  "resizeMode": "contain",
                },
                {
                  "backgroundColor": "#2222220D",
                  "borderColor": "#2222220D",
                  "borderRadius": 20,
                  "borderWidth": 0.5,
                  "height": 40,
                  "width": 40,
                },
              ]
            }
          />
        </View>
      </View>
      <View
        style={
          {
            "flexDirection": "column",
            "flexShrink": 1,
            "gap": 2,
          }
        }
      >
        <Text
          allowFontScaling={true}
          maxFontSizeMultiplier={1.4}
          numberOfLines={1}
          style={
            {
              "color": "#222222",
              "fontFamily": "Basel-Book",
              "fontSize": 19,
              "lineHeight": 24,
            }
          }
          suppressHighlighting={true}
        >
          cum
        </Text>
        <Text
          allowFontScaling={true}
          maxFontSizeMultiplier={1.4}
          numberOfLines={1}
          style={
            {
              "color": "#7D7D7D",
              "fontFamily": "Basel-Book",
              "fontSize": 17,
              "lineHeight": 24,
            }
          }
          suppressHighlighting={true}
          testID="token-item/metadata-subtitle"
        />
      </View>
      <View
        style={
          {
            "alignItems": "center",
            "flexDirection": "row",
            "flexGrow": 1,
            "justifyContent": "flex-end",
          }
        }
      >
        <View
          style={
            {
              "flexDirection": "row",
            }
          }
        >
          <View
            style={
              {
                "alignItems": "flex-end",
                "flexDirection": "column",
                "gap": 4,
                "minWidth": 70,
              }
            }
          >
            <Text
              allowFontScaling={true}
              maxFontSizeMultiplier={1.4}
              style={
                {
                  "color": "#222222",
                  "fontFamily": "Basel-Book",
                  "fontSize": 19,
                  "lineHeight": 24,
                }
              }
              suppressHighlighting={true}
              testID="token-item/price"
            >
              -
            </Text>
            <View
              style={
                {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "gap": 2,
                  "justifyContent": "flex-start",
                }
              }
              testID="relative-change"
            >
              <View
                style={
                  {
                    "flexDirection": "column",
                  }
                }
              >
                <Text
                  allowFontScaling={true}
                  maxFontSizeMultiplier={1.4}
                  style={
                    {
                      "color": "#7D7D7D",
                      "fontFamily": "Basel-Book",
                      "fontSize": 17,
                      "lineHeight": 24,
                    }
                  }
                  suppressHighlighting={true}
                >
                  -
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
